torch/_prims/__init__.py

```diff
@@ -1235,21 +1237,21 @@ def _greater_than_reduce(acc, x):
 
     # shape must be broadcastable to
     for idx, new_idx in enumerate(broadcast_dimensions):
-        assert a.shape[idx] == 1 or a.shape[idx] == shape[new_idx]
+        assert guard_size_oblivious(a.shape[idx] == 1) or a.shape[idx] == shape[new_idx]
```

Comment:
Making sure I am following along.

In the unbacked case we will assume that (`a.shape[idx]` == 1`) will give False because we will set it's bound to be 2. 

That will give us the `a.shape[idx] == shape[new_idx]` case which will of the unbacked symint being equal to the broadcasting dimension but because it's a size oblivious guard we won't guard. Which is what we want here.

However I think ideally for this case we would want something like:

`assert deferred_guard(a.shape[idx] == 1 | a.shape[idx] == shape[new_idx])`

which would allow us to know that the guard is either equal is either 1 or `shape[new_idx]` (and set the ranges appropriately)

Reply:
Yeah, the parallel or thing was my very first implementation a year ago. We can still do it, but it isn't actually getting you better code in the end, because we aren't generating guards on the unbacked SymInts anyway, so the main point is just getting past this assert without failing, which the current spelling also does.

Reply:
It does give you better ranges.. which might let you skip through the next guard which might not be size oblivious on the unbacked symint.

Actually, won't `guard_size_oblivious(a.shape[idx] == 1)`  eval as false and then won't we end up at `a.shape[idx] == shape[new_idx]` which isn't guard oblivious ? what am i missing there.

Reply:
I guess I do not understand concretely what range refinement you are hoping would happen here. Suppose you have `a.shape[idx] in [0, inf]` and `shape[new_idx] in [5, inf]`, are you hoping that a.shape[idx] will get an updated range? But it is impossible for this to occur, because if `a.shape[idx]` and `shape[new_idx]` are not obviously equal (e.g., `u0 == u0`), the second equality will also fail and you'll fail the assert.

It might be that you're looking for something like:

```
if not guard_size_oblivious(a.shape[idx] == 1):
    torch._check(a.shape[idx] == shape[new_idx])
```

This version will induce a runtime assert that `a.shape[idx] == shape[new_idx]`. There isn't any parallel or going on here, but I suppose this would allow us to unify in some cases where we previously couldn't. Is this the idea?